pyam/core.py

```diff
@@ -311,21 +311,25 @@ def as_pandas(self, with_metadata=False):
         """
         if with_metadata:
             if isinstance(with_metadata, dict):
-                cols = set(['exclude'])
-                for arg, value in with_metadata.items():
-                    if isstr(value) and value in self.meta.columns:
-                        cols.add(value)
+                cols = self._discover_meta_cols(**with_metadata)
             else:
                 cols = self.meta.columns
             return (
                 self.data
                 .set_index(META_IDX)
-                .join(self.meta[list(cols)])
+                .join(self.meta[cols])
                 .reset_index()
             )
         else:
             return self.data.copy()
 
+    def _discover_meta_cols(self, **kwargs):
```

Comment:
even thought its private, could you please add a docstring explaining what this function does?